File: server/assets/solutions/ohtu-avoin-2021/part6/python/querylanguage/src/querybuilder.py
from matchers import All, And, PlaysIn, HasAtLeast, Not, Or

class Querybuilder:
    def __init__(self, matcher = All()):
        self._matcher = matcher

    def build(self):
        return self._matcher

    def playsIn(self, team):
        return Querybuilder(And(self._matcher, PlaysIn(team)))

    def hasAtLeast(self,  value, attr):
        return Querybuilder(And(self._matcher, HasAtLeast(value, attr)))

    def hasFewerThan(self,  value, attr):
        return Querybuilder(And(self._matcher, Not(HasAtLeast(value, attr))))

    def oneOf(self, m1, m2):
        return Querybuilder(Or(m1, m2))

File: server/assets/solutions/ohtu-avoin-2020/part6/QueryLanguage/src/main/java/statistics/QueryBuilder.java

package statistics;

import statistics.matcher.All;
import statistics.matcher.And;
import statistics.matcher.HasAtLeast;
import statistics.matcher.HasFewerThan;
import statistics.matcher.Matcher;
import statistics.matcher.Or;
import statistics.matcher.PlaysIn;

public class QueryBuilder {
    private Matcher matcher;

    public QueryBuilder() {
        matcher = new All();
    }    
    
    public Matcher build() {
       Matcher  built = matcher; 
       matcher = new All();
       return built;
    } 

    public QueryBuilder playsIn(String team) {
        matcher = new PlaysIn(team);
        return this;
    }

    public QueryBuilder hasAtLeast(int number, String what) {
        matcher = new And(matcher, new HasAtLeast(number, what));
        return this;
    }
    
    public QueryBuilder hasFewerThan(int number, String what) {
        matcher = new And(matcher, new HasFewerThan(number, what));
        return this;
    }    

    public QueryBuilder oneOf(Matcher m1, Matcher m2) {
        matcher = new Or(m1, m2);
        return this;
    }
}
